### PR TITLE
Added OpenJPEG source code as a part of 3rdparty

### PR DESCRIPTION
**Pull request by VadimLevin**
OpenJPEG support for JPEG 2000 was added via #16494 and next step is to add OpenJPEG source code to 3rdparty as it is done for other libraries used by the imgcodecs module.

This patch adds OpenJPEG 2.3.1 to 3rdparty.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
allow_multiple_commits=1
```

### CODE REVIEW

**Comment by mshabunin** [comment]
I hope we would consider using modern methods of adding 3rdparty dependencies: 
* hthttps://cmake.org/cmake/help/v3.11/module/FetchContent.html
* https://cmake.org/cmake/help/v3.0/module/ExternalProject.html
* https://git-scm.com/docs/git-submodule

**Comment by alalek** [comment]
Current approach is fine for non-huge 3rdparty projects.

Modern methods require additional investigation:
- ExternalProject can't work with CMake targets
- FetchContent is just a downloader (and requires latest CMake, Ubuntu 18.04 contains CMake 3.10.x only)
- install rules?

---

@VadimLevin Please squash commits onto one, we don't need to keep whole history (at least removed files).

**Comment by mshabunin** [comment]
> FetchContent is just a downloader

Yes and we have our own downloader `ocv_download` which is used to [download and build TBB library](https://github.com/opencv/opencv/blob/master/3rdparty/tbb/CMakeLists.txt) for example.

**Comment by alalek** [comment]
@VadimLevin Please add temporary test commit with enabled option [here](https://github.com/opencv/opencv/blob/4.4.0/CMakeLists.txt#L209) (will be removed before merge through `git rebase`)

Please add `BUILD_OPENJPEG` option few lines below and add handling.

**Comment by VadimLevin** [comment]
> @VadimLevin Please add temporary test commit with enabled option [here](https://github.com/opencv/opencv/blob/4.4.0/CMakeLists.txt#L209) (will be removed before merge through `git rebase`)
> 
> Please add `BUILD_OPENJPEG` option few lines below and add handling.

Done (warnings are not related to OpenJPEG)

Unfortunately, build fails on Windows and it confuses me, because linker string looks correct and all libraries are built. 
```
/OUT:"C:\Users\testuser\Documents\OpenCV\opencv\build\bin\Release\opencv_imgcodecs440.dll" /MANIFEST /NXCOMPAT /PDB:"C:/Users/testuser/Documents/OpenCV/opencv/build/bin/Release/opencv_imgcodecs440.pdb" /DYNAMICBASE "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib" "..\..\lib\Release\opencv_imgproc440.lib" "..\..\3rdparty\lib\Release\ippiw.lib" "..\..\3rdparty\ippicv\ippicv_win\icv\lib\intel64\ippicvmt.lib" "..\..\3rdparty\lib\Release\zlib.lib" "..\..\3rdparty\lib\Release\libjpeg-turbo.lib" "..\..\3rdparty\lib\Release\libwebp.lib" "..\..\3rdparty\lib\Release\libpng.lib" "..\..\3rdparty\lib\Release\libtiff.lib" "..\..\3rdparty\lib\Release\openjp2.lib" "..\..\3rdparty\lib\Release\IlmImf.lib" "..\..\lib\Release\opencv_core440.lib" /IMPLIB:"C:/Users/testuser/Documents/OpenCV/opencv/build/lib/Release/opencv_imgcodecs440.lib" /DLL /MACHINE:X64 /NODEFAULTLIB:"atlthunk.lib" /NODEFAULTLIB:"atlsd.lib" /NODEFAULTLIB:"libcmt.lib" /INCREMENTAL:NO /PGD:"C:\Users\testuser\Documents\OpenCV\opencv\build\bin\Release\opencv_imgcodecs440.pgd" /SUBSYSTEM:CONSOLE /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /ManifestFile:"opencv_imgcodecs.dir\Release\opencv_imgcodecs440.dll.intermediate.manifest" /ERRORREPORT:PROMPT /NOLOGO /TLBID:1
```
It fails to find all symbols from `openjp2`, but it presents in the pointed directory.

**Comment by VadimLevin** [comment]
@alalek I fixed all configuration and build errors related to the OpenJPEG built from sources, can you proceed with review, please?

**Comment by VadimLevin** [comment]
@alalek I fixed all configuration and build errors related to the OpenJPEG built from sources, can you proceed with review, please?

File: 3rdparty/openjpeg/openjp2/j2k.c [hunk 1]
```diff
@@ -6447,7 +6447,7 @@ OPJ_BOOL opj_j2k_set_threads(opj_j2k_t *j2k, OPJ_UINT32 num_threads)
     return OPJ_FALSE;
 }
 
-static int opj_j2k_get_default_thread_count()
+static int opj_j2k_get_default_thread_count(void)
```
Comment (alalek) [inline 3rdparty/openjpeg/openjp2/j2k.c:6450]: Please do NOT modify 3rdparty sources.
Silence related warnings through `ocv_warnings_disable()` instead.

**Review by alalek — COMMENTED** [review]
Please squash all non-test commits into one to keep small patch (we don't need whole history with added/removed files).

**Comment by VadimLevin** [comment]
@alalek Thank you for review, I've reverted changes in 3rdparty code and suppressed warnings, everything should pass smoothly.

**Comment by alalek** [comment]
After cleanup of dead files:

> patch size opencv: 297 KiB 

I believe we can bypass `ocv_download()` approach here (size of `openjpeg-2.3.1.tar.gz` is 2Mb)

**Review by alalek — APPROVED** [review]
Thank you!

**Pull request merged**